FileAlreadyExistsException even when overwrite=true if skipCache=false

I'm using the configuration mentioned below, with `overwrite` set to true but the build fails with `FileAlreadyExistsException` if `skipCache` is false (by default false).

Setting `skipCache` to true explicitly alongwith `overwrite` to true makes it work, but then you have to look in the code to understand it.

Shouldn't overwrite handle this directly?
Let me know if you want me to submit a patch for this, I'll be happy to help.

**Goal used:** wget
**Plugin Version used:** 1.6.6

```
<plugin>
    <groupId>com.googlecode.maven-download-plugin</groupId>
    <artifactId>download-maven-plugin</artifactId>
    <version>1.6.6</version>
    <executions>
        <execution>
            <id>get-spark-sql-grammar</id>
            <phase>generate-sources</phase>
            <goals>
                <goal>wget</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <overwrite>true</overwrite>
        <!-- <skipCache>true</skipCache> -->
        <outputFileName>SqlBase.g4</outputFileName>
        <followRedirects>true</followRedirects>
        <url>
            https://raw.githubusercontent.com/apache/spark/v${spark.version}/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
        </url>
        <outputDirectory>${project.basedir}/src/main/antlr4/imports/</outputDirectory>
    </configuration>
</plugin>
```

```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.googlecode.maven-download-plugin:download-maven-plugin:1.6.6:wget (get-grammar) on project project: IO Error
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: IO Error
    at com.googlecode.download.maven.plugin.internal.WGet.execute (WGet.java:457)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.nio.file.FileAlreadyExistsException: /Users/chitralverma/IdeaProjects/project/core/src/main/antlr4/imports/test.g4
    at sun.nio.fs.UnixCopyFile.copy (UnixCopyFile.java:551)
    at sun.nio.fs.UnixFileSystemProvider.copy (UnixFileSystemProvider.java:253)
    at java.nio.file.Files.copy (Files.java:1274)
    at com.googlecode.download.maven.plugin.internal.WGet.execute (WGet.java:416)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

--- a/src/it/overwriteWithSkippedCache/invoke.properties
+++ b/src/it/overwriteWithSkippedCache/invoke.properties
@@ -1,0 +1,2 @@
+invoker.goals = clean verify
+invoker.buildResult = success

--- a/src/it/overwriteWithSkippedCache/pom.xml
+++ b/src/it/overwriteWithSkippedCache/pom.xml
@@ -1,0 +1,99 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+
+	<!-- Build description -->
+	<groupId>com.googlecode.maven-download-plugin.it</groupId>
+	<artifactId>testBasic</artifactId>
+	<packaging>pom</packaging>
+    <version>${testing.versionUnderTest}</version>
+	<name>Test</name>
+
+	<!-- Build plugins and extensions -->
+	<build>
+		<plugins>
+			<!--  Download the file first time -->
+			<plugin>
+                <groupId>@project.groupId@</groupId>
+                <artifactId>@project.artifactId@</artifactId>
+                <version>@project.version@</version>
+				<executions>
+					<execution>
+                        <id>first-download</id>
+						<phase>generate-resources</phase>
+						<goals>
+							<goal>wget</goal>
+						</goals>
+						<configuration>
+							<url>${test.img.file.url}</url>
+							<skipCache>true</skipCache>
+							<overwrite>true</overwrite>
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
+			<!--
+				Store last modification timestamp of downloaded file into 'stamp' file
+				to be compared on verification stage.
+			-->
+			<plugin>
+				<groupId>org.codehaus.gmavenplus</groupId>
+				<artifactId>gmavenplus-plugin</artifactId>
+                <version>1.5</version>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.codehaus.groovy</groupId>
+                        <artifactId>groovy</artifactId>
+                        <version>3.0.8</version>
+                        <scope>runtime</scope>
+                    </dependency>
+                </dependencies>
+				<executions>
+					<execution>
+						<id>save-timestamp</id>
+						<phase>process-resources</phase>
+						<goals>
+							<goal>execute</goal>
+						</goals>
+						<configuration>
+							<properties>
+								<target>${project.build.directory}</target>
+                                <testfilename>${test.img.file.name}</testfilename>
+							</properties>
+							<scripts>
+								<script>
+									<![CDATA[
+                                        File file = new File(target, testfilename);
+                                        File stamp = new File(target, "stamp");
+                                        stamp << file.lastModified();
+                                        assert true;
+                                    ]]>
+								</script>
+							</scripts>
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
+			<!--  Download the file second time -->
+            <plugin>
+                <groupId>@project.groupId@</groupId>
+                <artifactId>@project.artifactId@</artifactId>
+                <version>@project.version@</version>
+                <executions>
+                    <execution>
+                        <id>second-download</id>
+                        <phase>prepare-package</phase>
+                        <goals>
+                            <goal>wget</goal>
+                        </goals>
+                        <configuration>
+                            <url>${test.img.file.url}</url>
+                            <skipCache>false</skipCache>
+                            <overwrite>true</overwrite>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+		</plugins>
+	</build>
+
+</project>

--- a/src/it/overwriteWithSkippedCache/verify.groovy
+++ b/src/it/overwriteWithSkippedCache/verify.groovy
@@ -1,0 +1,21 @@
+/**
+ * Covers the bug https://github.com/maven-download-plugin/maven-download-plugin/issues/194
+ *
+ * Scenario:
+ * Given plugin configuration:
+ *  - skipCache = false (default)
+ *  - overwrite = true
+ *  - no checksum specified
+ * When Run the plugin to download the zip
+ * And Save timestamp into "stamp" file
+ * And Run the plugin to download the same zip
+ * Then the file should be overwritten
+ *
+ * `img_file_name` is injected through scriptVariables
+ */
+File target = new File(basedir, "target")
+File zip = new File(target, img_file_name)
+assert zip.exists()
+File stamp = new File(target, "stamp")
+assert stamp.exists() : "Timestamp file should exist"
+assert stamp.text != String.valueOf(zip.lastModified()) : "File expected to be modified but it wasn't"

--- a/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java
+++ b/src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java
@@ -397,15 +397,11 @@ public class WGet extends AbstractMojo {
                         checksumMatch = false;
                     }
                 }
-                if (!checksumMatch) {
+                if (!checksumMatch || overwrite) {
                     outputFile.delete();
                     haveFile = false;
-                } else if (!overwrite) {
-                    getLog().info("File already exist, skipping");
                 } else {
-                    // If no checksum provided and owerwriting requested we
-                    // will treat the fact as if there is no file in the cache.
-                    haveFile = false;
+                    getLog().info("File already exist, skipping");
                 }
             }
 

